File: nim/day04.nim
import strutils, sets, algorithm, sequtils

const passphrases = readFile("./inputs/04.txt").splitLines()

func areOnlyDistinct(containter: seq[string]): bool =
  len(containter) == len(containter.toHashSet)

func letterSort(word: string): string =
  sorted(word).join()

var
  first, second: int
  words, anagrams: seq[string]

for line in passphrases:
  words = line.split()
  anagrams = words.map(letterSort)
  if words.areOnlyDistinct: inc first
  if anagrams.areOnlyDistinct: inc second

echo first
echo second

File: python/day04.py
with open('./inputs/04.txt') as f:
    passphrases = f.readlines()

first = 0
second = 0

are_only_distinct = lambda x: len(x) == len(set(x))

for line in passphrases:
    words = line.split()
    first += are_only_distinct(words)
    anagrams = [''.join(sorted(word)) for word in words]
    second += are_only_distinct(anagrams)

print(first)
print(second)
